leader-election-impl/src/main/java/com/palantir/leader/PaxosLeaderElectionService.java

```diff
@@ -118,23 +120,87 @@ public PaxosLeaderElectionService(PaxosProposer proposer,
         this.proposer = proposer;
         this.knowledge = knowledge;
         // XXX This map uses something that may be proxied as a key! Be very careful if making a new map from this.
         this.otherPotentialLeadersToHosts = Collections.unmodifiableMap(otherPotentialLeadersToHosts);
         this.acceptors = ImmutableList.copyOf(acceptors);
         this.learners = ImmutableList.copyOf(learners);
-        this.executor = executor;
+        this.leaderPingExecutors = createLeaderPingExecutors(otherPotentialLeadersToHosts, executorServiceFactory);
+        this.knowledgeUpdatingExecutors = createKnowledgeUpdateExecutors(learners, executorServiceFactory);
         this.updatePollingRateInMs = updatePollingWaitInMs;
         this.randomWaitBeforeProposingLeadership = randomWaitBeforeProposingLeadership;
         this.leaderPingResponseWaitMs = leaderPingResponseWaitMs;
         lock = new ReentrantLock();
         this.eventRecorder = eventRecorder;
         this.latestRoundVerifier = new CoalescingPaxosLatestRoundVerifier(
-                new PaxosLatestRoundVerifierImpl(acceptors, proposer.getQuorumSize(), executor,
+                new PaxosLatestRoundVerifierImpl(
+                        acceptors,
+                        proposer.getQuorumSize(),
+                        createLatestRoundVerifierExecutors(acceptors, executorServiceFactory),
                         onlyLogOnQuorumFailure));
     }
 
+    private Map<PingableLeader, ExecutorService> createLeaderPingExecutors(
+            Map<PingableLeader, HostAndPort> otherLeadersToHosts,
+            Function<String, ExecutorService> executorServiceFactory) {
+        Map<PingableLeader, ExecutorService> executors = Maps.newHashMap();
+        executors.put(this, executorServiceFactory.apply("leader-ping-0"));
+
+        int index = 1;
+        for (PingableLeader leader : otherLeadersToHosts.keySet()) {
+            executors.put(leader, executorServiceFactory.apply("leader-ping-" + index));
+            index++;
+        }
+
+        return executors;
+    }
+
+    private Map<PaxosLearner, ExecutorService> createKnowledgeUpdateExecutors(
+            List<PaxosLearner> paxosLearners,
+            Function<String, ExecutorService> executorServiceFactory) {
+        return IntStream.range(0, paxosLearners.size())
+                .boxed()
+                .collect(Collectors.toMap(
+                        paxosLearners::get,
+                        index -> executorServiceFactory.apply("knowledge-update-" + index)));
+    }
+
+    private Map<PaxosAcceptor, ExecutorService> createLatestRoundVerifierExecutors(
+            List<PaxosAcceptor> paxosAcceptors,
+            Function<String, ExecutorService> executorServiceFactory) {
+        Map<PaxosAcceptor, ExecutorService> executors = Maps.newHashMap();
+        for (int i = 0; i < paxosAcceptors.size(); i++) {
+            executors.put(paxosAcceptors.get(i), executorServiceFactory.apply("lates-round-verifier-" + i));
```

Comment:
typo: `lates` -> `latest`

Reply:
it would be pretty nice for the `PaxosAcceptor` and `PaxosLearner` had uuid's similar to `PingableLeader`.